### ISSUE TITLE
How data should be stored on the network

### ISSUE THREAD

**Issue by xuset** [xuset/p2pweb#5]
This issue is for a discussion on how to store data on the network, and there are a lot of different considerations, like:
 * Should peers be homogeneous or should certain peers perform special functions?
 * There needs to be many copies of the same data stored in the network in order to prevent data loss
 * Also, some data loss is required. The total amount of data that is able to be stored is bounded (and constantly changing) so there must be some way to discard unwanted data
 * The total amount of data the network is able to store (at a given time) is constrained by a few factors like: The number of peers online, the average amount of data a peer is able to store. As well as the churn rate of peers, and the average transfer rate between peers.
 * How do you go from "I want data x" to "I found peers who have data x"? This is closely related to the "peer connection topology" issue #4.
   * Finding peers who have the data you want is a problem that is made more difficult with peers constantly coming and going
 * Ability to discover data. Ex: When a peer is loading a post-based site, it doesn't know exactly what posts to pull, it just wants the 10 most recent/popular posts.
 * Security is a good concern. Think about how a malicious person might try to break it

A good solution to how data should be stored on the network is needed. A discussion on how to approach this is welcome

**Comment by xuset** [comment]
The idea I have for this one is a hierarchical data store that supports static and dynamic content.

## The structure of the data store

Every piece of data, aka block, is identifiable by it's path in the hierarchy. Certain paths can be specified by the website operator to allow user generated content. For example, the `/post` path could be configured to allow users to push content to it. When a post is pushed to `/post` it is identified by it's parents path prepended to it's hash so the path of the post becomes `/post/<hash>`. This can be nested further like: `/post/<post_hash>/comments/<comment_hash>`.

The hierarchical approach allows peers to discover relevant data. For example, a peer could pull the 10 most recent or popular items in `/post` or pull all the comments for `/post/<hash>/comments`.

What's important to note is that the storage structure looks like a hierarchy of logs/arrays/whatever. These logs can have data (blocks) pushed to them and pulled from them, but the data itself is static. So logs are dynamic, and the data they hold is static.

## Where the data is stored

The idea I keep coming back to is: peers only seed the data they pull/push. This is similar to the bittorrent approach where you don't seed torrents you don't download. This allows the network to naturally discard data that is not used, old, or spammy. It also naturally supports many redundant copies of important (popular) data while less important data has fewer copies.

## Discovering data

As explained in the first section, data is discovered by pulling the items in a log. Once those items are pulled, the peer is now seeding them which is why what data is discovered is important. When pulling from `/post`, you won't want old posts, you will want new posts. So items in the logs should be ordered by age (or some other metric).

When a new item is pushed, it is pushed to the front of the log; so the newly pushed item is the first to be pulled. This allows new data to be kept around while data at the back of the log is naturally discarded when peers stop pulling it.

As a side not: using a universal ordering by age for logs is rigid. It would be better to allow applications to define their own order, and default to age if one is not supplied.

## Conclusion

What are your thoughts?

**Comment by georgeaf99** [comment]
I like the heirarchy structure, concept that data is immutable, and that clients should only seed posts that they view; however, I think we may be able to have implement the storage in a different way to make discovery easier. Each peer keeps a _directory_ of all posts without any specific information about individual posts. This _directory_ is versioned (either using a diff or vector clocks) and constantly replicated by peers in the system. Clients can get post content using a [discovery algorithm](www.cs.cmu.edu/~harchol/Papers/discovery.ps) or a lookup to a DHT.

**Comment by xuset** [comment]
In my idea, peers wouldn't seed only the posts they view, they would seed *all* the posts they pulled, which could be the 10,20, or 100 most recent/popular posts.

Our two ideas really aren't that different. The _directory_ is almost synonymous with the log. The main differences that I see are that logs are not versioned, and directories are not bounded in size. Also directories store keys (right?), and logs store the content

**Comment by xuset** [comment]
I would like to version data, but I can't come up with a good way to do it. How would diffs or vector clocks work?

**Comment by georgeaf99** [comment]
I agree that peers would seed all of the posts they have viewed in the past (up to some pre-determined limit).

My idea with the directory is that it contains every known post and each peer has the whole copy of it. We could version the directory by doing a diff and comparing the posts with the latest timestamp.

**Comment by xuset** [comment]
I like that idea that every peer knows about all the data in the network. It just needs a way to limit the amount of data that can be stored; the network doesn't have unlimited storage, and that must be taken into account

**Comment by georgeaf99** [comment]
If each entry in the directory consists of a post description and UUID/link they could probably be fit in 250 bytes. Let's say the directory size is trimmed down to one megabyte, then we can have 4,000 posts in the directory. I think that's probably enough for now?

**Comment by xuset** [comment]
The size of the directory will almost always be small, but the data stored in the network won't be. If only text data is stored in the post-based website example, it probably will be small enough to fit entirely on the network, but the lib should be general and not assume that all the data can fit on the network all the time.

**Comment by georgeaf99** [comment]
I agree; however, for now I think we can just tail the directory and accept the upper bound on maximum capacity in the network. In the future we can definitely expand on this system to make it better.
